### PR TITLE
quant: fix bug with copy.deepcopy of FX prepared quantization models

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46895 quant: fix bug with copy.deepcopy of FX prepared quantization models**

Summary:

Bug: models after the FX graph mode quant prepare step lost information,
such as the extra attributes defined in `Quantizer.save_state`,
if the user performed `copy.deepcopy` on them.  The information was lost
because `GraphModule` does not copy attributes which are not present on
`nn.Module` by default.

Fix: define a custom `__deepcopy__` method on observed models and
whitelist the attributes we care about.

This is needed because users sometimes run `copy.deepcopy` on their
models during non-quantization related preparations, and we should make
sure that quantization related state survives these calls.

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_deepcopy
python test/test_quantization.py TestQuantizeFx.test_standalone_module
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24556035](https://our.internmc.facebook.com/intern/diff/D24556035)

### CODE REVIEW

File: torch/quantization/fx/observed_module.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    # to copy the quantization specific attributes correctly.
+    def __deepcopy__(self, memo):
+        fake_mod = torch.nn.Module()
+        fake_mod.__dict__ = copy.deepcopy(self.__dict__)
+        return ObservedGraphModule(fake_mod, self.graph)
+
+def mark_observed_module(module):
+    return ObservedGraphModule(module, module.graph)
+
+class ObservedStandaloneGraphModule(ObservedGraphModule):
```
Comment (vkuzo) [inline torch/quantization/fx/observed_module.py:31]: this was moved from `standalone_module.py`, seemed cleaner to keep them in the same file

File: torch/quantization/fx/observed_module.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        for attr in preserved_attrs:
+            setattr(self, attr, preserved_attrs[attr])
+
+    # GraphModule does not copy attributes which are not in the __dict__
+    # of vanilla nn.Module.  So, we override __deepcopy__ in order
+    # to copy the quantization specific attributes correctly.
+    def __deepcopy__(self, memo):
+        fake_mod = torch.nn.Module()
+        fake_mod.__dict__ = copy.deepcopy(self.__dict__)
+        return ObservedGraphModule(fake_mod, self.graph)
```
Comment (vkuzo) [inline torch/quantization/fx/observed_module.py:26]: if there is a way in Python to instantiate the current class without hardcoding the class name, we could make this function fully reusable as well.  Didn't find it after a 5 minute search, but welcome to suggestions

File: torch/quantization/fx/quantize.py [hunk 3]
```diff
@@ -496,6 +497,7 @@ def is_observed(input_arg):
 
         model = GraphModule(model, observed_graph)
```
Comment (vkuzo) [inline torch/quantization/fx/quantize.py:498]: following the existing guidance, but seems like directly instantiating `ObservedGraphModule` might also make sense here

File: torch/quantization/fx/quantize.py [hunk 4]
```diff
@@ -26,7 +26,8 @@
     get_default_quant_patterns,
 )
 
-from .standalone_module import (
+from .observed_module import (
+    mark_observed_module,
```
Comment (jerryzh168) [inline torch/quantization/fx/quantize.py:30]: I think we can also define an `is_observed_module` and remove some of the validations in https://github.com/pytorch/pytorch/blob/master/torch/quantization/fx/quantize.py#L516-L522

**Review by jerryzh168 — APPROVED** [review]
Looks good, thanks! had a nit comment on removing some validation code.

**Pull request closed without merging**